test/boost/lister_test.cc

```diff
@@ -0,0 +1,183 @@
+    for (size_t i = 0; i < count; i++) {
+        auto name = tests::random::get_sstring(tests::random::get_int(1, 8));
+        if (tests::random::get_bool()) {
+            if (!file_names.contains(name)) {
+                dir_names.insert(name);
+            }
+        } else {
+            if (!dir_names.contains(name)) {
+                file_names.insert(name);
+            }
+        }
+    }
+
+    for (const auto& name : file_names) {
+        co_await touch_file(tmp.path() / name);
+    }
+    for (const auto& name : dir_names) {
+        co_await touch_directory((tmp.path() / name).native());
+    }
+
+    co_return file_names.size() + dir_names.size();
+}
+
+SEASTAR_TEST_CASE(test_queue_lister) {
+    auto tmp = tmpdir();
+
+    std::unordered_set<std::string> file_names;
+    std::unordered_set<std::string> dir_names;
+
+    auto count = co_await generate_random_content(tmp, file_names, dir_names);
+    BOOST_TEST_MESSAGE(fmt::format("Generated {} dir entries", count));
+
+    std::unordered_set<std::string> found_file_names;
+    std::unordered_set<std::string> found_dir_names;
+
+    auto ql = queue_lister(tmp.path());
+
+    while (auto de = co_await ql.get()) {
+        assert(de->type);
+        switch (*de->type) {
+        case directory_entry_type::regular: {
+            auto [it, inserted] = found_file_names.insert(de->name);
+            assert(inserted);
+            break;
+        }
+        case directory_entry_type::directory: {
+            auto [it, inserted] = found_dir_names.insert(de->name);
+            assert(inserted);
+            break;
+        }
+        default:
+            BOOST_FAIL(fmt::format("Unexpected directory_entry_type: {}", *de->type));
+        }
+    }
+
+    BOOST_REQUIRE(found_file_names == file_names);
+    BOOST_REQUIRE(found_dir_names == dir_names);
+}
+
+SEASTAR_TEST_CASE(test_queue_close) {
+    auto tmp = tmpdir();
+
+    std::unordered_set<std::string> file_names;
+    std::unordered_set<std::string> dir_names;
+
+    auto count = co_await generate_random_content(tmp, file_names, dir_names, tests::random::get_int(100, 1000));
+    BOOST_TEST_MESSAGE(fmt::format("Generated {} dir entries", count));
+
+    {
+        auto ql = queue_lister(tmp.path());
+        auto initial = tests::random::get_int(count);
+        BOOST_TEST_MESSAGE(fmt::format("Getting {} dir entries", initial));
+        for (auto i = 0; i < initial; i++) {
+            auto de = co_await ql.get();
+            assert(de);
+        }
+        BOOST_TEST_MESSAGE("Closing queue_lister");
+        co_await ql.close();
+    }
+
+    {
+        auto ql = queue_lister(tmp.path());
+        auto initial = tests::random::get_int(count);
+        BOOST_TEST_MESSAGE(fmt::format("Getting {} dir entries", initial));
+        for (auto i = 0; i < initial; i++) {
+            auto de = co_await ql.get();
+            assert(de);
+        }
+        BOOST_TEST_MESSAGE("Closing queue_lister");
+        co_await ql.close();
```

Comment:
Unless I'm missing something, this block is an **exact** copy of the previous block - so why do it twice? You can do the extra `ql.get()` check that you do below, after the first block - no need for the second block at all.

Reply:
the first one destroys the `queue_lister` after `close()`,
this one, after the extra `get()` call.

Maybe I can extract the inner-body of the block into a worker function...

Reply:
But who cares about this difference? the extra get() call just fails and you check that - and then you can destroy. Why do you need both with and without the get? I don't see what it checks additionally.

Reply:
I want to make sure that the queue_lister can be destroyed safely right after close, possibly with stuff in the queue, vs. destroyed after `get()` returns an error.  It needs to ensure that the underlying open directory is closed properly (and by the way, it's not when the walker throws an exception IIRC - needs fixing)